File: mobile-sdk/kotlin/ApiClient.kt
package com.yourapp.api

import io.github.jan.supabase.createSupabaseClient
import io.github.jan.supabase.gotrue.Auth
import io.github.jan.supabase.gotrue.auth
import io.github.jan.supabase.postgrest.Postgrest
import io.github.jan.supabase.realtime.Realtime
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import org.json.JSONObject

/**
 * API Client for Delivery Hub
 *
 * Usage:
 * val apiClient = ApiClient(SUPABASE_URL, SUPABASE_ANON_KEY)
 * apiClient.initialize()
 */
class ApiClient(
    private val supabaseUrl: String,
    private val supabaseKey: String
) {
    private val supabase = createSupabaseClient(supabaseUrl, supabaseKey) {
        install(Auth)
        install(Postgrest)
        install(Realtime)
    }

    private val functionsUrl = "$supabaseUrl/functions/v1"
    private var heartbeatJob: Job? = null

    /**
     * Get current auth token
     */
    private suspend fun getAuthToken(): String? {
        return supabase.auth.currentSessionOrNull()?.accessToken
    }

    /**
     * Make authenticated API call
     */
    private suspend fun apiCall(
        endpoint: String,
        body: JSONObject,
        requiresAuth: Boolean = true
    ): Result<JSONObject> {
        return try {
            val token = if (requiresAuth) getAuthToken() else null
            if (requiresAuth && token == null) {
                return Result.failure(Exception("Not authenticated"))
            }

            // Make HTTP call here using your preferred HTTP library (OkHttp, Ktor, etc.)
            // This is a placeholder - implement actual HTTP call

            Result.success(JSONObject())
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    /**
     * Start heartbeat - call every 5 minutes
     */
    fun startHeartbeat() {
        heartbeatJob?.cancel()
        heartbeatJob = CoroutineScope(Dispatchers.IO).launch {
            while (isActive) {
                try {
                    val body = JSONObject().apply {
                        put("app_type", "mobile")
                    }
                    apiCall("heartbeat", body)
                } catch (e: Exception) {
                    // Log error but don't crash
                }
                delay(5 * 60 * 1000) // 5 minutes
            }
        }
    }

    /**
     * Stop heartbeat when app goes to background
     */
    fun stopHeartbeat() {
        heartbeatJob?.cancel()
        heartbeatJob = null
    }

    /**
     * Log error to server
     */
    suspend fun logError(
        errorMessage: String,
        stackTrace: String?,
        appVersion: String,
        deviceInfo: Map<String, Any>
    ): Result<Unit> {
        val body = JSONObject().apply {
            put("error_message", errorMessage)
            put("stack_trace", stackTrace)
            put("app_version", appVersion)
            put("device_info", JSONObject(deviceInfo))
        }

        return apiCall("log-error", body).map { }
    }

    /**
     * Report issue
     */
    suspend fun reportIssue(
        description: String,
        appVersion: String
    ): Result<Unit> {
        val body = JSONObject().apply {
            put("description", description)
            put("app_version", appVersion)
        }

        return apiCall("report-issue", body).map { }
    }

    /**
     * Check app version
     */
    suspend fun checkAppVersion(
        currentVersion: String,
        platform: String = "android"
    ): Result<VersionCheckResponse> {
        val body = JSONObject().apply {
            put("current_version", currentVersion)
            put("platform", platform)
        }

        return apiCall("check-app-version", body, requiresAuth = false)
            .map { json ->
                VersionCheckResponse(
                    latestVersion = json.getString("latest_version"),
                    currentVersion = json.getString("current_version"),
                    needsUpdate = json.getBoolean("needs_update"),
                    forceUpdate = json.getBoolean("force_update"),
                    isDeprecated = json.getBoolean("is_deprecated"),
                    downloadUrl = json.optString("download_url"),
                    releaseNotes = json.optString("release_notes")
                )
            }
    }

    /**
     * Update job status
     */
    suspend fun updateJobStatus(
        jobId: Long,
        status: JobStatus
    ): Result<Unit> {
        val body = JSONObject().apply {
            put("job_id", jobId)
            put("status", status.value)
        }

        return apiCall("update-job-status", body).map { }
    }

    /**
     * Get optimized route
     */
    suspend fun getOptimizedRoute(
        pickup: String,
        delivery: String
    ): Result<RouteResponse> {
        val body = JSONObject().apply {
            put("pickup", pickup)
            put("delivery", delivery)
        }

        return apiCall("get-optimized-route", body)
            .map { json ->
                RouteResponse(
                    polyline = json.getString("polyline"),
                    distance = json.getJSONObject("distance"),
                    duration = json.getJSONObject("duration")
                )
            }
    }

    /**
     * Track analytics event
     */
    suspend fun trackAnalytics(
        eventName: String,
        eventData: Map<String, Any>? = null,
        appVersion: String,
        platform: String = "android",
        sessionId: String
    ): Result<Unit> {
        val body = JSONObject().apply {
            put("event_name", eventName)
            put("event_data", JSONObject(eventData ?: emptyMap()))
            put("app_version", appVersion)
            put("platform", platform)
            put("session_id", sessionId)
        }

        return apiCall("track-analytics", body).map { }
    }
}

// Data classes
data class VersionCheckResponse(
    val latestVersion: String,
    val currentVersion: String,
    val needsUpdate: Boolean,
    val forceUpdate: Boolean,
    val isDeprecated: Boolean,
    val downloadUrl: String,
    val releaseNotes: String
)

data class RouteResponse(
    val polyline: String,
    val distance: JSONObject,
    val duration: JSONObject
)

enum class JobStatus(val value: String) {
    IN_PROGRESS("in_progress"),
    COMPLETED("completed"),
    CANCELLED("cancelled")
}

File: mobile-sdk/typescript/api-client.ts
/**
 * TypeScript/React Native API Client
 *
 * Install dependencies:
 * npm install @supabase/supabase-js
 */

import { createClient, SupabaseClient } from '@supabase/supabase-js'

export interface DeviceInfo {
  os: string
  os_version: string
  device_model: string
  app_build: string
}

export interface VersionCheckResponse {
  latest_version: string
  current_version: string
  needs_update: boolean
  force_update: boolean
  is_deprecated: boolean
  download_url?: string
  release_notes?: string
}

export interface RouteResponse {
  polyline: string
  distance: {
    text: string
    value: number
  }
  duration: {
    text: string
    value: number
  }
}

export class ApiClient {
  private supabase: SupabaseClient
  private functionsUrl: string
  private heartbeatInterval?: NodeJS.Timeout

  constructor(supabaseUrl: string, supabaseKey: string) {
    this.supabase = createClient(supabaseUrl, supabaseKey)
    this.functionsUrl = `${supabaseUrl}/functions/v1`
  }

  /**
   * Get auth token
   */
  private async getAuthToken(): Promise<string | null> {
    const { data } = await this.supabase.auth.getSession()
    return data.session?.access_token || null
  }

  /**
   * Make authenticated API call
   */
  private async apiCall<T = any>(
    endpoint: string,
    body: any,
    requiresAuth = true
  ): Promise<T> {
    const token = requiresAuth ? await this.getAuthToken() : null
    if (requiresAuth && !token) {
      throw new Error('Not authenticated')
    }

    const headers: Record<string, string> = {
      'Content-Type': 'application/json',
    }

    if (token) {
      headers['Authorization'] = `Bearer ${token}`
    }

    const response = await fetch(`${this.functionsUrl}/${endpoint}`, {
      method: 'POST',
      headers,
      body: JSON.stringify(body),
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.error || 'API call failed')
    }

    return response.json()
  }

  /**
   * Start heartbeat
   */
  startHeartbeat() {
    this.stopHeartbeat() // Clear any existing interval

    this.heartbeatInterval = setInterval(async () => {
      try {
        await this.apiCall('heartbeat', { app_type: 'mobile' })
      } catch (error) {
        console.error('Heartbeat failed:', error)
      }
    }, 5 * 60 * 1000) // 5 minutes

    // Send initial heartbeat immediately
    this.apiCall('heartbeat', { app_type: 'mobile' }).catch(console.error)
  }

  /**
   * Stop heartbeat
   */
  stopHeartbeat() {
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval)
      this.heartbeatInterval = undefined
    }
  }

  /**
   * Log error
   */
  async logError(
    errorMessage: string,
    stackTrace: string | null,
    appVersion: string,
    deviceInfo: DeviceInfo
  ): Promise<void> {
    await this.apiCall('log-error', {
      error_message: errorMessage,
      stack_trace: stackTrace,
      app_version: appVersion,
      device_info: deviceInfo,
    })
  }

  /**
   * Report issue
   */
  async reportIssue(description: string, appVersion: string): Promise<void> {
    await this.apiCall('report-issue', {
      description,
      app_version: appVersion,
    })
  }

  /**
   * Check app version
   */
  async checkAppVersion(
    currentVersion: string,
    platform: 'android' | 'ios' = 'android'
  ): Promise<VersionCheckResponse> {
    return this.apiCall<VersionCheckResponse>(
      'check-app-version',
      {
        current_version: currentVersion,
        platform,
      },
      false
    )
  }

  /**
   * Update job status
   */
  async updateJobStatus(
    jobId: number,
    status: 'in_progress' | 'completed' | 'cancelled'
  ): Promise<void> {
    await this.apiCall('update-job-status', {
      job_id: jobId,
      status,
    })
  }

  /**
   * Get optimized route
   */
  async getOptimizedRoute(
    pickup: string,
    delivery: string
  ): Promise<RouteResponse> {
    return this.apiCall<RouteResponse>('get-optimized-route', {
      pickup,
      delivery,
    })
  }

  /**
   * Track analytics
   */
  async trackAnalytics(
    eventName: string,
    eventData?: Record<string, any>,
    appVersion?: string,
    platform?: string,
    sessionId?: string
  ): Promise<void> {
    await this.apiCall('track-analytics', {
      event_name: eventName,
      event_data: eventData || {},
      app_version: appVersion,
      platform,
      session_id: sessionId,
    })
  }

  /**
   * Sign in
   */
  async signIn(email: string, password: string) {
    return this.supabase.auth.signInWithPassword({ email, password })
  }

  /**
   * Sign out
   */
  async signOut() {
    this.stopHeartbeat()
    return this.supabase.auth.signOut()
  }

  /**
   * Get current user
   */
  async getCurrentUser() {
    const { data } = await this.supabase.auth.getUser()
    return data.user
  }
}

// Usage example:
// const apiClient = new ApiClient(SUPABASE_URL, SUPABASE_ANON_KEY)
// apiClient.startHeartbeat()
//
// // When app goes to background:
// apiClient.stopHeartbeat()
//
// // When app returns to foreground:
// apiClient.startHeartbeat()
